### DISCUSSION TITLE
Is there a way to modify the parse tree similar to pylint

### DISCUSSION THREAD

**Discussion by jesse-sony** [microsoft/pyright#3516] [Q&A]
We have a custom import management system for our tools. We've looked at converting it to work with python import hooks instead, but haven't gotten to that. So in our code, an import might look like

`importTool.importPackage("packageName")`
after which you can use `packageName` just as you would something imported via
`import packageName`
Obviously this completely breaks any sort of import resolution using pyright. With pylint, we were able to write a plugin that allowed us to modify the AST and replace these statements with import symbols so that pylint can correctly parse all our imports. I was wondering if there was a way to do something similar in pyright or if some sort of generic parser modification might be considered as a feature.

cheers
jesse c

**Comment by erictraut** [comment]
Pyright doesn't support plugins and doesn't allow this sort of customization. We don't have plans to support anything like this. If you want to support this, you'd need to fork the pyright code and create a private version.

An alternative is to use a `TYPE_CHECKING` conditional in your code so pyright sees standard import statements whereas the runtime executes your dynamic import resolution calls. If and when you implement a python import hook, you could then switch the runtime over to use these import statements directly.
